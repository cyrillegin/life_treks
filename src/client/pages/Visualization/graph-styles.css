.chart-axis-shape {
  fill: none;
  shape-rendering: crispEdges;
  stroke: #000;
  stroke-width: 1px;
}

.chart-axis-shape > g > line {
  stroke-opacity: 0.2;
}

.chart-axis-shape > g > text {
  stroke: none;
  font-family: 'Roboto', 'Helvetica', 'Arial', sans-serif;
}
