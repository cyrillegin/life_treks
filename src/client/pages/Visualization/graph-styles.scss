$black: #000;

.chart-axis-shape {
  fill: none;
  shape-rendering: crispEdges;
  stroke: $black;
  stroke-width: 1px;
}

.chart-axis-shape > g > line {
  stroke-opacity: 0.2;
}

.chart-axis-shape > g > text {
  font-family: 'Roboto', 'Helvetica', 'Arial', sans-serif;
  stroke: none;
}
