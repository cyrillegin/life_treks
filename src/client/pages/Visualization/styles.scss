$red: #f00;

.root {
  margin: 16px;
}

.paper {
  padding: 20px;
  text-align: center;
}

.graph-title {
  text-align: center;
}

.stats-contianer {
  flex-grow: 1;
}

.input-container {
  flex-grow: 1;
}

.description-container {
  flex-grow: 1;
  margin-left: 12px;
  margin-right: 20px;
  max-width: 50%;
  text-align: justify;
}

.extras {
  display: flex;
  text-align: left;
}

.list {
  list-style-type: none;
}

.error-message {
  color: $red;
}
