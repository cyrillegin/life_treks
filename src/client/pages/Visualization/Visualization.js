import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {withStyles} from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import Graph from './graph';

const styles = theme => ({
  root: {
    margin: '16px',
  },
  paper: {
    padding: '20px',
    textAlign: 'center',
  },
  graphTitle: {
    marginLeft: '8px',
    textAlign: 'left',
  },
  spinner: {
    margin: 'auto',
  },
  timeControls: {
    textAlign: 'right',
    flexGrow: '1',
    display: 'inline-block',
    margin: '10px',
  },
  textField: {
    margin: '12px',
  },
  statsContianer: {
    flexGrow: 1,
  },
  inputContainer: {
    flexGrow: 1,
  },
  extras: {
    display: 'flex',
  },
  list: {
    listStyleType: 'none',
  },
});

export class VisualizationPage extends Component {

  static propTypes = {
    classes: PropTypes.object.isRequired,
  }

  state = {
    readings: [],
    stats: {
      min: 0,
      max: 0,
      avg: 0,
      last: 0,
      count: 0,
    },
    equation: 'x ^ 2',
    points: [],
    xMin: -10,
    xMax: 10,
    yMin: 0,
    yMax: 20,
  }

  parseEquation() {
    const arr = this.state.equation.split(' ').join('');
    let tree = {
      left: null,
      right: null,
      value: null,
    };
    for (let i = 0; i < arr.length; i++) {
      let value = arr[i];
      switch (arr[i]) {
        case '+':
        case '-':
        case '*':
        case '/':
        case '^':
          tree = {
            left: tree,
            value: value,
            right: null,
          };
          break;
        default:
          if (!isNaN(parseInt(value))) {
            let cursor = 1;
            while (!isNaN(parseInt(arr[i + cursor]))) {
              value = value + arr[i + cursor];
              cursor++;
            }
            i += cursor - 1;
          }
          if (tree.value === null) {
            tree.value = value;
          } else if (tree.left === null) {
            tree.left = {
              left: null,
              right: null,
              value: value,
            };
          } else if (tree.right === null) {
            tree.right = {
              left: null,
              right: null,
              value: value,
            };
          }
      }
    }
    const newSet = [];
    for (let i = this.state.xMin; i <= this.state.xMax; i++) {
      newSet.push({
        x: i,
        y: this.getValue(tree, i),
      });
    }

    this.setState({
      points: newSet,
    });
  }

  getValue(tree, value) {
    // leaf node
    if (tree.left === null && tree.right === null) {
      if (isNaN(parseInt(tree.value))) {
        return value;
      }
      return parseInt(tree.value);
    }

    // operator
    switch (tree.value) {
      case '+':
        return this.getValue(tree.left, value) + this.getValue(tree.right, value);
      case '-':
        return this.getValue(tree.left, value) - this.getValue(tree.right, value);
      case '*':
        return this.getValue(tree.left, value) * this.getValue(tree.right, value);
      case '/':
        return this.getValue(tree.left, value) / this.getValue(tree.right, value);
      case '^':
        return Math.pow(this.getValue(tree.left, value), this.getValue(tree.right, value));
    }
  }

  componentDidMount() {
    this.parseEquation();
  }

  render() {

    const submitEquation = () => {
      this.parseEquation();
    };

    const handleEqInput = e => event => {
      this.setState({
        equation: event.target.value,
      });
    };

    const handleRangeInput = e => event => {
      let value = 0;
      if (!isNaN(parseInt(event.target.value))) {
        value = parseInt(event.target.value);
      }

      this.setState({
        [e]: value,
      }, this.parseEquation);

    };

    return (
      <div className={this.props.classes.root}>
        <Paper className={this.props.classes.paper} elevation={4}>
          <Typography variant="headline" component="h3" className={this.props.classes.graphTitle}>
            graph
          </Typography>
          <Graph
            points={this.state.points}
            ranges={{
              xMin: this.state.xMin,
              xMax: this.state.xMax,
              yMin: this.state.yMin,
              yMax: this.state.yMax,
            }}
          />
          <div className={this.props.classes.extras}>
            <div className={this.props.classes.statsContianer}>
              <table>
                <tbody>
                  <tr>
                    <td>Min: </td>
                    <td>{this.state.stats.min}</td>
                  </tr>
                  <tr>
                    <td>Max:</td>
                    <td>{this.state.stats.max}</td>
                  </tr>
                  <tr>
                    <td>Avg:</td>
                    <td>{this.state.stats.avg}</td>
                  </tr>
                  <tr>
                    <td>Last:</td>
                    <td> {this.state.stats.last}</td>
                  </tr>
                  <tr>
                    <td>Count:</td>
                    <td> {this.state.stats.count}</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div className={this.props.classes.inputContainer}>
              Enter your equation:
              <input type="text" onChange={handleEqInput('name')} />
              <br />
              X Minimum:
              <input type="text" onChange={handleRangeInput('xMin')} />
              <br />
              X Maximum
              <input type="text" onChange={handleRangeInput('xMax')} />
              <br />
              Y Minimum
              <input type="text" onChange={handleRangeInput('yMin')} />
              <br />
              Y Maximum
              <input type="text" onChange={handleRangeInput('yMax')} />
              <br />
              <button onClick={submitEquation}>
                Submit
              </button>

            </div>
          </div>
        </Paper>
      </div>
    );
  }
}

export default withStyles(styles)(VisualizationPage);
