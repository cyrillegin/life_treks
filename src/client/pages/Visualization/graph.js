import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {withStyles} from '@material-ui/core/styles';
import './graph-styles.css';

const d3 = {
  ...require('d3-selection'), // eslint-disable-line
  ...require('d3-scale'), // eslint-disable-line
  ...require('d3-axis'), // eslint-disable-line
  ...require('d3-format'), // eslint-disable-line
  ...require('d3-shape'), // eslint-disable-line
  ...require('d3-array') // eslint-disable-line
};

const styles = theme => ({
  root: {},
});

export class Graph extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    points: PropTypes.arrayOf(PropTypes.shape({
      x: PropTypes.number.isRequired,
      y: PropTypes.number.isRequired,
    })).isRequired,
    ranges: PropTypes.shape({
      xMin: PropTypes.number.isRequired,
      xMax: PropTypes.number.isRequired,
      yMin: PropTypes.number.isRequired,
      yMax: PropTypes.number.isRequired,
    }).isRequired,
  }

  state = {
    mounted: false,
  }

  drawGraph() {
    console.log(this.props);
    if (this.props.points.length === 0) {
      return;
    }
    // Initialization.
    const container = document.querySelector('#graph-container');
    container.innerHTML = '';

    let width = container.clientWidth;

    let height = 400;
    const margin = {
      top: 10,
      right: 20,
      bottom: 10,
      left: 30,
    };
    width = width - margin.left - margin.right;
    height = height - margin.top - margin.bottom;

    // Create the svg.
    const newChart = d3.select('#graph-container')
      .append('svg')
      .attr('width', width + margin.left + margin.right)
      .attr('height', height + margin.top + margin.bottom)
      .append('g')
      .attr('transform', `translate(${margin.left},${margin.bottom})`)
      .classed('svg-content-responsive', true);


    const xScale = d3.scaleLinear()
      .domain([this.props.ranges.xMin, this.props.ranges.xMax]) // input
      .range([0, width]); // output

    const yScale = d3.scaleLinear()
      .domain([this.props.ranges.yMin, this.props.ranges.yMax]) // input
      .range([height - margin.bottom, margin.top]); // output


    // Y Axis
    function getTic(yMin, yMax) {
      const Ticks = [];
      const ratio = (yMax - yMin) / 6;
      for (let q = 0; q < 7; q ++) {
        Ticks.push(yMin + (ratio * q));
      }
      return Ticks;
    }

    const yAxis = d3.axisLeft(yScale)
      .tickSizeInner(- width)
      .tickSizeOuter(0)
      .tickValues(getTic(this.props.ranges.yMax, this.props.ranges.yMin));

    newChart.append('g')
      .attr('class', 'chart-axis-shape')
      .call(yAxis);

    // X Axis
    const xAxis = d3.axisBottom(xScale)
      .tickSizeInner(- height + margin.bottom + margin.top)
      .tickSizeOuter(0)
      .tickPadding(10)
      .ticks(10);

    newChart.append('g')
      .attr('class', 'chart-axis-shape')
      .attr('transform', `translate(0,${height - margin.bottom})`)
      .call(xAxis);

    // Top border
    newChart.append('g')
      .append('line')
      .attr('class', 'chart-axis-shape')
      .attr('x1', 0)
      .attr('x2', width)
      .attr('y1', margin.bottom)
      .attr('y2', margin.bottom);

    // Right border
    newChart.append('g')
      .append('line')
      .attr('class', 'chart-axis-shape')
      .attr('x1', width)
      .attr('x2', width)
      .attr('y1', margin.bottom)
      .attr('y2', height - margin.bottom);

    // Graph lines
    const lineFunction = d3.line()
      .defined((d) => {
        return d.x <= this.props.ranges.xMax &&
          d.x >= this.props.ranges.xMin &&
          d.y >= this.props.ranges.yMin &&
          d.y <= this.props.ranges.yMax;
      })
      .x((d) => xScale(d.x))
      .y((d) => yScale(d.y));

    newChart.append('path')
      .attr('d', lineFunction(this.props.points))
      .attr('stroke', 'rgba(45, 100, 255, 0.87)')
      .attr('stroke-width', 2)
      .attr('fill', 'none');

  } // End D3

  componentDidMount() {
    this.drawGraph();
    this.setState({
      mounted: true,
    });
  }

  render() {
    if (this.state.mounted) {
      this.drawGraph();
    }
    return (
      <div className={this.props.classes.root}>
        <div id="graph-container" />
      </div>
    );
  }
}

export default withStyles(styles)(Graph);
