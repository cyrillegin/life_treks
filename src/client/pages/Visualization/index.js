import React, { Component } from 'react';
import Graph from './graph';
import './styles.scss';

export default class VisualizationPage extends Component {
  constructor() {
    super();
    this.state = {
      equation: 'x ^ 2',
      points: [],
      xMin: -10,
      xMax: 10,
      yMin: 0,
      yMax: 20,
      resoultion: 1,
      hasError: false,
    };
  }

  componentDidMount() {
    this.parseEquation();
  }

  checkInputs() {
    let err = false;
    if (isNaN(parseFloat(this.state.xMin))) {
      err = true;
    }
    if (isNaN(parseFloat(this.state.xMax))) {
      err = true;
    }
    if (isNaN(parseFloat(this.state.yMin))) {
      err = true;
    }
    if (isNaN(parseFloat(this.state.yMax))) {
      err = true;
    }
    if (isNaN(parseFloat(this.state.resoultion))) {
      err = true;
    }
    this.setState({
      hasError: err,
    });
    return !err;
  }

  parseEquation() {
    if (!this.checkInputs()) {
      return;
    }
    const arr = this.state.equation.split(' ').join('');
    let tree = {
      left: null,
      right: null,
      value: null,
    };
    for (let i = 0; i < arr.length; i++) {
      let value = arr[i];
      switch (arr[i]) {
        case '+':
        case '-':
        case '*':
        case '/':
        case '^':
          tree = {
            left: tree,
            value,
            right: null,
          };
          break;
        default:
          if (!isNaN(parseInt(value, 10))) {
            let cursor = 1;
            while (!isNaN(parseInt(arr[i + cursor], 10))) {
              value += arr[i + cursor];
              cursor++;
            }
            i += cursor - 1;
          }
          if (tree.value === null) {
            tree.value = value;
          } else if (tree.left === null) {
            tree.left = {
              left: null,
              right: null,
              value,
            };
          } else if (tree.right === null) {
            tree.right = {
              left: null,
              right: null,
              value,
            };
          }
      }
    }
    const newSet = [];
    try {
      for (
        let i = parseFloat(this.state.xMin);
        i <= parseFloat(this.state.xMax);
        i += parseFloat(1 / this.state.resoultion)
      ) {
        newSet.push({
          x: i,
          y: this.getValue(tree, i),
        });
      }
    } catch (e) {
      this.setState({
        hasError: true,
      });
    }

    this.setState({
      points: newSet,
    });
  }

  getValue = (tree, value) => {
    // leaf node
    if (tree.left === null && tree.right === null) {
      if (isNaN(parseInt(tree.value, 10))) {
        return value;
      }
      return parseInt(tree.value, 10);
    }

    // operator
    switch (tree.value) {
      case '+':
        return this.getValue(tree.left, value) + this.getValue(tree.right, value);
      case '-':
        return this.getValue(tree.left, value) - this.getValue(tree.right, value);
      case '*':
        return this.getValue(tree.left, value) * this.getValue(tree.right, value);
      case '/':
        return this.getValue(tree.left, value) / this.getValue(tree.right, value);
      case '^':
        return this.getValue(tree.left, value) ** this.getValue(tree.right, value);
      default:
        return '';
    }
  };

  render() {
    const handleEqInput = e => event => {
      this.setState(
        {
          equation: event.target.value,
        },
        this.parseEquation,
      );
    };

    const handleRangeInput = e => event => {
      this.setState(
        {
          [e]: event.target.value,
        },
        this.parseEquation,
      );
    };

    return (
      <div className="root">
        <div className="paper">
          <div className="graph-title">Graphing Calculator</div>
          <Graph
            points={this.state.points}
            ranges={{
              xMin: this.state.xMin,
              xMax: this.state.xMax,
              yMin: this.state.yMin,
              yMax: this.state.yMax,
            }}
          />
          {this.state.hasError && (
            <div className="error-message">Could not understand input parameters.</div>
          )}
          <div className="extras">
            <div className="input-container">
              <table>
                <tbody>
                  <tr>
                    <td>Enter your equation:</td>
                    <td>
                      <input
                        type="text"
                        value={this.state.equation}
                        onChange={handleEqInput('name')}
                        aria-label="Equation input"
                      />
                    </td>
                  </tr>
                  <tr>
                    <td>X Minimum:</td>
                    <td>
                      <input
                        type="text"
                        value={this.state.xMin}
                        onChange={handleRangeInput('xMin')}
                        aria-label="X axis minimum"
                      />
                    </td>
                  </tr>
                  <tr>
                    <td>X Maximum</td>
                    <td>
                      <input
                        type="text"
                        value={this.state.xMax}
                        onChange={handleRangeInput('xMax')}
                        aria-label="X axis maximum"
                      />
                    </td>
                  </tr>
                  <tr>
                    <td>Y Minimum</td>
                    <td>
                      <input
                        type="text"
                        value={this.state.yMin}
                        onChange={handleRangeInput('yMin')}
                        aria-label="Y axis minimum"
                      />
                    </td>
                  </tr>
                  <tr>
                    <td>Y Maximum</td>
                    <td>
                      <input
                        type="text"
                        value={this.state.yMax}
                        onChange={handleRangeInput('yMax')}
                        aria-label="Y axis Maximum"
                      />
                    </td>
                  </tr>
                  <tr>
                    <td>Resoultion</td>
                    <td>
                      <input
                        type="text"
                        value={this.state.resoultion}
                        onChange={handleRangeInput('resoultion')}
                        aria-label="Resoultion of the graph line"
                      />
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>

            <div className="description-container">
              This is an example of some custom graphing using the D3 library. You can enter an
              equation on the left side and set the dimensions and resoultion of the graph. The
              calculator is pretty basic as I made the parser from scratch. So far it only supports
              +, -, *, /, ^, and does not support order of operations.
            </div>
          </div>
        </div>
      </div>
    );
  }
}
