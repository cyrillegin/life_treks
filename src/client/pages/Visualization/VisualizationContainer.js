import {compose, mapProps} from 'recompose';
import Visualization from './Visualization';

export default compose(
  mapProps((ownProps) => {
    return {
      ...ownProps,
    };
  }),
)(Visualization);
