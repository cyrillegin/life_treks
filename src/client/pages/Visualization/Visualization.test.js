import React from 'react';
import renderer from 'react-test-renderer';
import Visualization from './Visualization';

describe('Visualization', () => {
  it('should render snapshots', () => {
    global.document.body.innerHTML = '<div id="graph-container"></div>';
    const tree = renderer.create(<Visualization classes={{}} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
