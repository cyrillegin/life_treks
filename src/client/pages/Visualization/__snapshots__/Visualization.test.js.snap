// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visualization should render snapshots 1`] = `
<div
  className="VisualizationPage-root-1"
>
  <div
    className="MuiPaper-root-10 MuiPaper-elevation4-16 MuiPaper-rounded-11 VisualizationPage-paper-2"
  >
    <h3
      className="MuiTypography-root-37 MuiTypography-headline-42 VisualizationPage-graphTitle-3"
    >
      Graphing Calculator
    </h3>
    <div
      className="Graph-root-73"
    >
      <div
        id="graph-container"
      />
    </div>
    <div
      className="VisualizationPage-extras-7"
    >
      <div
        className="VisualizationPage-inputContainer-5"
      >
        <table>
          <tbody>
            <tr>
              <td>
                Enter your equation:
              </td>
              <td>
                <input
                  aria-label="Equation input"
                  onChange={[Function]}
                  type="text"
                  value="x ^ 2"
                />
              </td>
            </tr>
            <tr>
              <td>
                X Minimum:
              </td>
              <td>
                <input
                  aria-label="X axis minimum"
                  onChange={[Function]}
                  type="text"
                  value={-10}
                />
              </td>
            </tr>
            <tr>
              <td>
                X Maximum
              </td>
              <td>
                <input
                  aria-label="X axis maximum"
                  onChange={[Function]}
                  type="text"
                  value={10}
                />
              </td>
            </tr>
            <tr>
              <td>
                Y Minimum
              </td>
              <td>
                <input
                  aria-label="Y axis minimum"
                  onChange={[Function]}
                  type="text"
                  value={0}
                />
              </td>
            </tr>
            <tr>
              <td>
                Y Maximum
              </td>
              <td>
                <input
                  aria-label="Y axis Maximum"
                  onChange={[Function]}
                  type="text"
                  value={20}
                />
              </td>
            </tr>
            <tr>
              <td>
                Resoultion
              </td>
              <td>
                <input
                  aria-label="Resoultion of the graph line"
                  onChange={[Function]}
                  type="text"
                  value={1}
                />
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        className="VisualizationPage-descriptionContainer-6"
      >
        This is an example of some custom graphing using the D3 library. You can enter an equation on the left side and set the dimensions and resoultion of the graph. The calculator is pretty basic as I made the parser from scratch. So far it only supports +, -, *, /, ^, and does not support order of operations.
      </div>
    </div>
  </div>
</div>
`;
